pkg/services/ngalert/api/authorization.go

```diff
@@ -216,79 +216,101 @@ func authorizeDatasourceAccessForRule(rule *ngmodels.AlertRule, evaluator func(e
 	return true
 }
 
+// authorizeAccessToRuleGroup checks all rules against authorizeDatasourceAccessForRule and exits on the first negative result
+func authorizeAccessToRuleGroup(rules []*ngmodels.AlertRule, evaluator func(evaluator ac.Evaluator) bool) bool {
+	for _, rule := range rules {
+		if !authorizeDatasourceAccessForRule(rule, evaluator) {
+			return false
+		}
+	}
+	return true
+}
+
 // authorizeRuleChanges analyzes changes in the rule group, and checks whether the changes are authorized.
 // NOTE: if there are rules for deletion, and the user does not have access to data sources that a rule uses, the rule is removed from the list.
 // If the user is not authorized to perform the changes the function returns ErrAuthorization with a description of what action is not authorized.
 // Return changes that the user is authorized to perform or ErrAuthorization
-func authorizeRuleChanges(change *changes, evaluator func(evaluator ac.Evaluator) bool) (*changes, error) {
-	var result = &changes{
-		GroupKey: change.GroupKey,
-		New:      change.New,
-		Update:   change.Update,
-		Delete:   change.Delete,
+func authorizeRuleChanges(change *changes, evaluator func(evaluator ac.Evaluator) bool) error {
+	namespaceScope := dashboards.ScopeFoldersProvider.GetResourceScopeUID(change.GroupKey.NamespaceUID)
+
+	affectedGroupsPermissions := make(map[ngmodels.AlertRuleGroupKey]bool, len(change.AffectedGroups))
+	for key, rules := range change.AffectedGroups {
+		hasAccess := true
```

Comment:
What about defaulting to false and requiring positive confirmation of access? Something like:
```go
authConfirmed := 0
for _, rule := range rules {
  if !authorized() {
    break;
  }
  authConfirmed++ 
}

if len(authConfirmed) != len(rules) {
  return ErrAuthorization
}
```

I think it protects us a bit if we write a buggy conditional in the future.

Reply:
Oh, thanks for pointing. I was going to replace it to authorizeAccessToRuleGroup which performs the same operation